public/app/core/components/NestedFolderPicker/NestedFolderPicker.tsx

```diff
@@ -385,3 +396,153 @@ const getStyles = (theme: GrafanaTheme2) => {
     }),
   };
 };
+
+// const listFoldersSelector = createSelector(
+//   (parentUid: string | undefined, page: number) => ({ parentUid, page }),
+//   (args) => browseDashboardsAPI.endpoints.listFolders.select(args)
+// );
+
+// export type DashboardViewItemCollection = {
+//   items: DashboardViewItem[];
+//   lastFetchedKind: 'folder' | 'dashboard';
+//   lastFetchedPage: number;
+//   lastKindHasMoreItems: boolean;
+//   isFullyLoaded: boolean;
+// };
+
+const listFolderSelector = createSelector(
+  [
+    (state: any) => state,
+    (state, parentUid: ListFolderArgs['parentUid']) => parentUid,
+    (state, parentUid: ListFolderArgs['parentUid'], page: ListFolderArgs['page']) => page,
+    (state, parentUid: ListFolderArgs['parentUid'], page: ListFolderArgs['page'], limit: ListFolderArgs['limit']) =>
+      limit,
+  ],
+  (state, parentUid, page, limit) => {
+    return browseDashboardsAPI.endpoints.listFolders.select({ parentUid, page, limit })(state);
+  }
+);
+
+function useFolderList() {
+  const requestedArgs = useRef<ListFolderArgs[] | null>(null);
+  if (requestedArgs.current === null) {
+    requestedArgs.current = [];
+  }
+
+  // TODO: propertly memoize this into a stable selector
+  const state = useSelector((rxState) => {
+    const requests = requestedArgs.current ?? [];
+
+    const pages = requests
+      .map((args) => {
+        return listFolderSelector(rxState, args.parentUid, args.page, args.limit);
+      })
+      .filter((page, index, all) => {
+        return all.findIndex((otherPage) => otherPage.requestId === page.requestId) === index;
+      });
+
+    const rootPages: Array<(typeof pages)[number]> = [];
+    const pagesByParent: Record<string, Array<(typeof pages)[number]>> = {};
+
+    for (const page of pages) {
+      const parentUid = page.originalArgs?.parentUid;
+
+      if (parentUid) {
+        if (!pagesByParent[parentUid]) {
+          pagesByParent[parentUid] = [];
+        }
+
+        pagesByParent[parentUid].push(page);
+      } else {
+        rootPages.push(page);
+      }
+    }
+
+    return {
+      rootPages,
+      pagesByParent,
+    };
+  });
+
+  console.log('state', state);
+
+  const rootCollection = useMemo(() => {
+    const rrr = state.rootPages;
+
+    let flatTree: Array<DashboardsTreeItem<DashboardViewItemWithUIItems>> = rrr.flatMap((page) => {
+      return (page.data ?? []).map((item) => {
+        const ddddd: DashboardsTreeItem<DashboardViewItemWithUIItems> = {
```

Comment:
Step 3, 4, ... 100: Then construct the flat tree from that state we selected out.

I think for I will re-create the createFlatTree function just for this component, but I've thought a little about how we could generalise it through selector functions.